File: Java/249-GroupShiftedStrings.java
public class Solution {
    public List<List<String>> groupStrings(String[] strings) {
        List<List<String>> res = new ArrayList<>();
        if (strings == null || strings.length == 0){
            return res;
        }
        Map<List<Integer>, List<String>> map = new HashMap<>();
        for (int i = 0; i < strings.length; i++){
            List<Integer> diff = new ArrayList<>();
            if (strings[i].length() > 1){
                for (int j = 1; j < strings[i].length(); j++){
                    int offset = strings[i].charAt(j) - strings[i].charAt(j - 1);
                    if (offset < 0){
                        offset += 26;
                    }
                    diff.add(offset);
                }
            }
            if (!map.containsKey(diff)){
                List<String> tmp = new ArrayList<>();
                tmp.add(strings[i]);
                map.put(diff, tmp);
            }else{
                List<String> tmp = map.get(diff);
                tmp.add(strings[i]);
                map.put(diff, tmp);
            }
        }
        return new ArrayList<>(map.values());
    }
}

File: cpp/249-GroupShiftedStrings.cpp
class Solution {
public:
    vector<vector<string>> groupStrings(vector<string>& strings) {
        unordered_map<string, vector<string>> umap;
        vector<vector<string>> res;
        for (int i = 0; i < strings.size(); i++) {
            string tmp = strings[i];
            int beginning = tmp[0] - 'a';
            //attention! the for loop
            for (int j = 0; j < tmp.size(); j++) {
                tmp[j] -= beginning;
                if (tmp[j] - 'a' < 0) {
                    tmp[j] += 26;
                }
            }
            umap[tmp].push_back(strings[i]);
        }
        for (auto pair : umap) {
            res.push_back(pair.second);
        }
        return res;
    }
};
